### PR TITLE
MBS-9479 (II): Show errors for dupe RG failures

### PR DESCRIPTION
**Pull request by derat**
# MBS-9479 (II)

Make the release editor display a "Failed loading existing release groups" error and retry link when a request to load possibly-duplicate release groups fails.

# Problem

#3359 made the release editor query for release groups with names similar to the release's, but it didn't display an error if the request failed due to network or server issues.

# Solution

Display a "Failed loading existing release groups." message and a "Retry" link:

Screenshot 2024-08-27 12 11 54

I moved the lookup code from `init.js` to `actions.js` so it can be called from the retry link.

# Testing

Manually tested using my local dev server. (I think I'm unable to test the no-error path now, since I think I can't make cross-origin calls to the `/ws/js/release-group` endpoint due to CORS restrictions.)

### CODE REVIEW

File: root/release/edit/information.tt [hunk 1]
```diff
@@ -51,33 +51,43 @@
         </td>
       </tr>
 
-      <tr data-bind="if: $root.loadingDuplicateRGs()">
+      <tr data-bind="if: $root.loadingDuplicateReleaseGroups()">
```
Comment (derat) [inline root/release/edit/information.tt:54]: I thought my earlier mix of "ReleaseGroup" and "RG" felt ugly, so I'm using "ReleaseGroup" everywhere now. Let me know if these feel too wordy, though.

File: root/static/scripts/release-editor/actions.js [hunk 2]
```diff
@@ -69,6 +75,83 @@ const actions = {
     }));
   },
 
+  duplicateReleaseGroups: ko.observableArray([]),
```
Comment (derat) [inline root/static/scripts/release-editor/actions.js:78]: I moved these to where they're set and used, but please let me know if keeping them in `init.js` would be preferable.

File: root/static/scripts/release-editor/actions.js [hunk 3]
```diff
@@ -69,6 +75,83 @@ const actions = {
...
+  loadingDuplicateReleaseGroups: ko.observable(false),
+  failedLoadingDuplicateReleaseGroups: ko.observable(false),
+  duplicateReleaseGroupsRequest: null,
```
Comment (derat) [inline root/static/scripts/release-editor/actions.js:81]: I tried to make this property observable and get rid of `loadingDuplicateReleaseGroups`, but I noticed strange behavior where modifying the observable `duplicateReleaseGroupsRequest` from within `loadDuplicateReleaseGroups` seemed to make my `debounceComputed(utils.withRelease(...))` get called again immediately.

File: root/static/scripts/release-editor/actions.js [hunk 4]
```diff
@@ -69,6 +75,83 @@ const actions = {
...
+  duplicateReleaseGroupsQuery: null,
+
+  /**
+   * Query for existing release groups that are credited to any of the
+   * currently-credited artists and have similar names to the release.
+   */
+  loadDuplicateReleaseGroups() {
```
Comment (derat) [inline root/static/scripts/release-editor/actions.js:88]: This is mostly just moved from `init.js`, but I'll try to point out the minor changes (apart from the calls to the new `failedLoadingDuplicateReleaseGroups` property) in review comments.

File: root/static/scripts/release-editor/actions.js [hunk 5]
```diff
@@ -69,6 +75,83 @@ const actions = {
...
+      releasegroup: [utils.escapeLuceneValue(release.name() ?? '')],
+    });
+    if (query === this.duplicateReleaseGroupsQuery) {
+      return;
+    }
+
+    // Cancel any in-progress lookup and clear existing results.
+    this.duplicateReleaseGroupsRequest?.abort();
+    this.duplicateReleaseGroupsRequest = null;
+    this.loadingDuplicateReleaseGroups(false);
```
Comment (derat) [inline root/static/scripts/release-editor/actions.js:103]: I wasn't updating this here before, but I think it makes sense to do so since the request is also being aborted and cleared.

File: root/static/scripts/release-editor/actions.js [hunk 6]
```diff
@@ -69,6 +75,83 @@ const actions = {
...
+
+    this.loadingDuplicateReleaseGroups(true);
+    this.duplicateReleaseGroupsRequest = request({
+      url: '/ws/js/release-group' +
+        '?direct=false' +
+        '&advanced=true' +
+        '&limit=' + encodeURIComponent(maxDuplicateReleaseGroups) +
+        '&q=' + encodeURIComponent(query),
+    })
+      .always(() => {
```
Comment (derat) [inline root/static/scripts/release-editor/actions.js:128]: I had to change the indenting here to avoid what seemed like it might be an ESLint bug:

```
  136:7  warning  Expected line break before `.done`  @stylistic/js/newline-per-chained-call
```

I think the indenting I was using that triggered the warning was like:

```js
this.duplicateReleaseGroupsRequest = request({
  // ...
}).always(() => {
  // ...
}).fail(() => {
  // ...
}).done((data) => {
  // ...
});
```

**Comment by derat** [comment]
@mwiencek and @reosarevok, mind taking a look?

@Aerozol, UX feedback is welcome. I tried various combinations of colors and bolding but didn't come up with anything that I love. I'm not sure if there are any existing places in the server where error messages are shown alongside retry links.

**Comment by Aerozol** [comment]
I think you can put it in one of these boxes:

image

Stretch goal would be to right-align the retry text/button, instead of the (x).
And unbold the text, I think the box will make it pretty strong.
I haven't had time to mock it up sorry, so might not look right or may need tweaking!

"Failed to load..." could read better?

**Comment by derat** [comment]
Here's what it looks like after using devtools to switch it to use the banner CSS:

Screenshot 2024-08-27 18 57 24

One downside of using a banner is that it'll shift the rest of the page content down due to being much taller than the loading message. This might make people misclick if they've already moved on to a lower part of the page.

Were you suggesting using the orange text color as well, or something else?  (Just to mention it, the orange isn't great from an a11y perspective -- it has a 2.81 : 1 contrast ratio against the gray banner background and 2.96 : 1 against white. [WCAG says it should be at least 4.5 : 1.](https://dequeuniversity.com/rules/axe/4.9/color-contrast))

**Comment by reosarevok** [comment]
@brainzbot, retest this please

**Review by reosarevok — COMMENTED** [review]
Tested locally and it works nicely - including retrying when I do give it access to the server. Not approving since I don't have clear answers to some of the questions, but if @mwiencek is happy, fine with merging this.

I personally think it looks just fine now and it would be significantly *worse* with the pictured banner style, but I'm not a designer :)

**Comment by Aerozol** [comment]
Definitely no orange text please! If it's going to shift everything then also not the box, thanks for checking that.

Maybe de-bold the text but make the search box go light red as well?

**Comment by derat** [comment]
> Maybe de-bold the text but make the search box go light red as well?

Sorry, but I suspect that that's going to be confusing. :-( The RG autocomplete box is a separate control that the user hasn't interacted with yet at this point, so I think it'll feel strange to put it into an error state if this background request fails. Plus, it may make the user think that they need to do something (e.g. selecting an existing RG) to resolve the issue, even though it's actually fine if they just go ahead at this point -- a new RG will be created like usual.

**Comment by Aerozol** [comment]
My apologies!! Too little sleep, too much work. I completely failed to closely look at what this does. I thought it was connected to the search box.

I think non-bold red or black text is fine, maybe italics, where it is.
Perhaps with the text: "Release group search error" or even just "Search error".
My reasoning is that the current text may be interpreted as a massive problem, rather than being able to continue (the combination of red + bold + "Failed to...")

**Comment by derat** [comment]
Here are a few permutations:

A
Screenshot 2024-08-28 18 31 00
B
Screenshot 2024-08-28 18 31 12
C
Screenshot 2024-08-28 18 31 19
D
Screenshot 2024-08-28 18 31 42
E
Screenshot 2024-08-28 18 34 28
F
Screenshot 2024-08-28 18 34 34

**Comment by Aerozol** [comment]
Thanks! I think italic, black and 'Release group search error' is tidy and non-heart attack inducing (not bold).

Red text if we need it to be stronger - but if this is more of an optional 'assistant', rather than a key feature, I don't think it should be red, which implies that things are really borked.

**Comment by reosarevok** [comment]
I'm fine with red or black here, and non-bolding it seems to be reasonable.

**Comment by derat** [comment]
Sure, here's a screenshot:

image

Hopefully editors will see this rarely enough that it doesn't really matter how it looks. :-)

**Comment by derat** [comment]
@mwiencek, does this look okay to you?

@brainzbot, retest this please

**Review by mwiencek — APPROVED** [review]
Works great, thanks!
